### PR TITLE
Cargo Automation

### PR DESCRIPTION
**Pull request by west3436**
PR adds 4 new machines to help automate common activities in the cargo bay:

https://user-images.githubusercontent.com/66280799/163265401-db71bd2e-b757-46f4-9c2b-827a7a338fe8.mp4

**Crate Opener**
- Opens crates that pass over it.
- Can open locked crates if a proper ID is swiped into the machine beforehand.
- Forces open locked crates if emagged.

**Crate Closer**
- Closes crates that pass over it.

**Item Sorting Machine**
- Variant of the destination/recycling sorting machine.
- Configure outputs via multitool and smack the machine with the item you want sorted out.

**Wrapping Machine**
- Wraps compatible items and affixes a destination tag, if chosen.
- Set outputs via multitool and destination tag via interacting with the machine.
- Can wrap people if loaded with Syndicate wrapping paper.
- Now featuring an improved sprite from @Blithering:

:cl:
 * rscadd: added magnetic crate open and closing machines.
 * rscadd: added wrapping and tagging machines for crates and wrappable items.
 * rscadd: added item sorting variant of the sorting machine.

### CODE REVIEW

File: code/game/objects/structures/crates_lockers/crates.dm [hunk 1]
```diff
@@ -525,6 +525,16 @@
 	else
 		to_chat(user, "<span class='notice'>Access Denied.</span>")
 
+/obj/structure/closet/crate/secure/proc/autotogglelock(var/obj/machinery/M)
```
Comment (jknpj) [inline code/game/objects/structures/crates_lockers/crates.dm]: you could just tweak togglelock() to avoid this spaghetti

**Comment by Eneocho** [comment]
cargobros... we won

**Comment by WindowsErrors** [comment]
> **Crate Opener**
> Can open locked crates if a proper ID is swiped into the machine beforehand.

i'm somewhat reminded of The Escapists where you have to make a key mold from a bar of soap if you want keys since if you just steal the keys without putting them back then the guards get all autistic and you get beaten and sent to solitary once the guard that the key was stolen from wakes up without their keys

**Comment by Cartlord** [comment]
this literally just makes cargo technicians do less
they already do so little, at this point remove the job

**Comment by west3436** [comment]
> this literally just makes cargo technicians do less they already do so little, at this point remove the job

shut up fartlord

**Comment by WindowsErrors** [comment]
they do have a point, this seems more like something I'd see on goon station than here

that's still not gonna stop the cargo player base from wanting it though

**Comment by west3436** [comment]
> they do have a point, this seems more like something I'd see on goon station than here
> 
> that's still not gonna stop the cargo player base from wanting it though

To address this meme argument seriously, none of these machines are going to be mapped in anywhere. If these ever are used in the cargo bay it'll be because the techs or QM themselves wanted to set it up. This will provide for more fun, not remove it.

**Comment by WindowsErrors** [comment]
hopefully your work doesn't get forgotten in that case

I'm already hard pressed to find the cargo workers utilizing the RCD or the kitchen staff using the hydroponics smartfridge or anyone using two thirds of the stuff in the research asteroid station that gets slept on constantly

File: code/game/objects/structures/crates_lockers/crates.dm (hunk 1 above)
Comment (west3436) [reply code/game/objects/structures/crates_lockers/crates.dm]: Implemented in latest commit

**Comment by Eneocho** [comment]
>or the kitchen staff using the hydroponics smartfridge

They always use it if there was a botanist making plants, since that's where you drop the harvests for the guy.

**Comment by RedrobotAnnie** [comment]
I love that it uses the same model as the borgification machine, now when I put it in cargo as a way to boost my CPU people will roll with it even more

**Comment by ManopolisJones** [comment]
why?

**Comment by WindowsErrors** [comment]
> why?

they saw it in a dream and felt a holy urge to bring it to our physical earthrealm

**Comment by Kurfursten** [comment]
I like this, not as a roundstart thing so I'm glad you clarified that, but this is a fun little thing cargo techies can make as they pimp their bay

We all know that playing cargo is 90% base building autism because people just don't constantly order crates. This is like the conveyorbelt medbay thing. It won't get done most shifts but it's there for players who want to take their cargo to the next level.

> re: unlocker

This is actually extremely strong, for what Windows said, but also because on revs/blob someone with a high access ID doesn't need to babysit cargo, you can just mass unlock crates. Since most shifts this machine won't be made, I actually don't think WE's worry is a major one, since you would need to really have the machine built in advance to do this quickly/covertly. Still, it's something we might want to think about and keep an eye on.

Space Law on Theft does include a section on opening crates inappropriately.

**Comment by tuypo1** [comment]
>Can open locked crates if a proper ID is swiped into the machine beforehand.

I think this would be better if the swipe only gave access for 10 minutes
emag being permanent is fine though

**Comment by ghost** [comment]
these are gonna be ones you can build and place later, right?

**Comment by west3436** [comment]
> these are gonna be ones you can build and place later, right?

dude did you scroll up at all

**Comment by WindowsErrors** [comment]
i never brought anything up as a negative thing or as a worry, i think the only issue with the crate opener is probably the rather inconspicuous sprite it currently has (you can barely see the lights from under the conveyer belt)

what i am more interested about is where you'd actually get the circuitry for all this machinery

soldering iron simple circuit? autolathe mathematics assemblies that nobody's used since 2017? hydroponics crops? trader crate? cargo crate? hydroponics crop? space or derelict or asteroid loot? m*chanics, or even r*search (6 gorillon trash shipments sent and counting)?

**Comment by Kurfursten** [comment]
It looks like he's putting them in the circuit printer, which feels reasonable. I would also be fine with it if he made the circuitboards orderable as a crate from cargo.

Your suggestion to get more obvious sprites is a good one though.

**Comment by west3436** [comment]
Yeah I was thinking about making the crates orderable as well - I will do so.
As for a sprite, I have put out requests for sprites as I am not a spriter by any means. I don't want that to preclude merging this if possible as these sprites "work", but I'd be happy to make a PR to update them whenever they are provided to me.

**Comment by Kurfursten** [comment]
Keep an eye on coder discord sprite channel, there is some interest from Blithering to help here.

edit: and boxed

**Comment by pedr69** [comment]
> cargo workers utilizing the RCD

what
you mean the rapid crate sender?
i've used it once to try and send 50 emitters to engineering and gave up because the thing is SO FUCKING SLOW

**Comment by luigil101** [comment]
combine this with auto-harvesting material plants and you can get some truly powerful factorio gimmicks for making infinite money
would need an auto crate builder for that though

**Comment by Exxion** [comment]
I have no problem with these as constructible machines as they are in this PR but I can't wait to complain on the PR in like five years to add them roundstart

**Comment by adacovsk** [comment]
> dude did you scroll up at all

write it in the dang PR, I was just about to downvote until I read the comments

**Comment by Viciousaurus** [comment]
If the crate opener doesn't disembowel people when emagged then it's useless

**Pull request merged**